walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,122 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+  public ArrayList<TimeRange> getMergedTimes(ArrayList<TimeRange> takenTimes){
+    Collections.sort(takenTimes, TimeRange.ORDER_BY_START);
+    TimeRange previous = takenTimes.get(0);
+    ArrayList<TimeRange> takenTimesMerged = new ArrayList<>();
+    for(TimeRange time: takenTimes){
+        if(time.overlaps(previous)){
+            int start = previous.start();
+            int end = Math.max(time.end(),previous.end());
+            previous = TimeRange.fromStartEnd(start,end,false);
+        }
+        else{
+            takenTimesMerged.add(previous);
+            previous = time;
+        }
+    }
+    takenTimesMerged.add(previous);
+    return takenTimesMerged;
+  }
+
+  public ArrayList<TimeRange> getBetweenTimes(ArrayList<TimeRange> takenTimesMerged){
+    ArrayList<TimeRange> validTimes = new ArrayList<>();
+    TimeRange firstValid = takenTimesMerged.get(0);
+    TimeRange lastValid = takenTimesMerged.get(takenTimesMerged.size() - 1);
+    int first = firstValid.start();
+    int last = lastValid.end();
+    validTimes.add(TimeRange.fromStartEnd(0,first,false));
+    validTimes.add(TimeRange.fromStartEnd(last,1440,false));
+
+    if(takenTimesMerged.size()>1){
+        for (int i = 0; i < takenTimesMerged.size(); i++) {
+            if(i == takenTimesMerged.size()-1){
+                break;
+            }
+            TimeRange currentRange = takenTimesMerged.get(i);
+            TimeRange adjRange = takenTimesMerged.get(i+1);
+            int newStart = currentRange.end();
+            int newEnd = adjRange.start();
+            validTimes.add(TimeRange.fromStartEnd(newStart,newEnd,false));
+        }
+    }
+    return validTimes;
   }
+
+  public ArrayList<TimeRange> filterTimes(ArrayList<TimeRange> validTimes, MeetingRequest request){
+    ArrayList<TimeRange> finalTimes = new ArrayList<>();
+    for(TimeRange time : validTimes){
+        if(time.duration() >= request.getDuration()){
+            finalTimes.add(time);
+        }
+    }
+    Collections.sort(finalTimes, TimeRange.ORDER_BY_START);
+    return finalTimes;
+  }
+
```

Comment:
"query" is a really broad word that doesn't give me much of a hint as to how I'm expected to use this method without digging into the source code. This would be an excellent place to add an explanatory comment about what the method does and what the individual params are expected to be.